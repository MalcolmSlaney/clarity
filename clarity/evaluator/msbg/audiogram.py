"""Supplied dataclass to represent a monaural audiogram"""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Final

import numpy as np
from scipy.interpolate import interp1d

if TYPE_CHECKING:
    from numpy import ndarray

DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES: Final = (
    250,
    500,
    1000,
    2000,
    3000,
    4000,
    6000,
    8000,
)

FULL_STANDARD_AUDIOGRAM_FREQUENCIES: Final = np.array(
    [
        125,
        250,
        500,
        750,
        1000,
        1500,
        2000,
        3000,
        4000,
        6000,
        8000,
        10000,
        12000,
        14000,
        16000,
    ]
)


def logx_interp1d(x_in: ndarray, y_in: ndarray, x_out: ndarray) -> ndarray:
    """Linear interpolation on logarithmic x-axis.

    Interpolates linearly on a logarithmic x-axis, e.g. suitable for
    interpolating audiogram levels at arbitrary frequencies.

    Note, for x below the minimum of x_in, the first value of y_in is used,
    and for x above the maximum of x_in, the last value of y_in is used.

    Args:
        x_in (ndarray): x-axis values, e.g. audiogram frequencies (Hz)
        y_in (ndarray): y-axis values, e.g., audiogram levels
        x_out (ndarray): x-axis value at which y-value are required

    Returns:
        ndarray: interpolated y-values

    """
    return interp1d(
        np.log(x_in), y_in, bounds_error=False, fill_value=(y_in[0], y_in[-1])
    )(np.log(x_out))


@dataclass
class Audiogram:
    """Dataclass to represent an audiogram.

    Attributes:
        levels (ndarray): The levels for the left and right ear
        frequencies (ndarray): The frequencies at which the levels are measured
    """

    levels: np.ndarray
    frequencies: np.ndarray = np.array(DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES)

    def __post_init__(self) -> None:
        """Check that dimensions of levels and frequencies match."""
        if len(self.levels) != len(self.frequencies):
            raise ValueError(
                f"Levels ({len(self.levels)}) and frequencies ({len(self.frequencies)})"
                " must have the same length"
            )

    @property
    def severity(self) -> str:
        """Categorise HL severity level for the audiogram.

        Note that this categorisation is different from that of the British
        Society of Audiology, which recommends descriptors mild, moderate,
        severe and profound for average hearing threshold levels at 250, 500,
        1000, 2000 and 4000 Hz of 21-40 dB HL, 41-70 dB HL, 71-95 dB HL
        and > 95 dB HL, respectively (BSA Pure-tone air-conduction and
        bone-conduction threshold audiometry with and without masking
        2018).

        Returns:
            str -- severity level, one of SEVERE, MODERATE, MILD, NOTHING

        """
        # calculate mean hearing loss between critical frequencies of 2 & 8 kHz
        critical_freqs = np.logical_and(
            2000 <= self.frequencies, self.frequencies <= 8000
        )
        critical_levels = self.levels[critical_freqs]
        # Remove any None values
        critical_levels = [x for x in critical_levels if x is not None]
        # Ignore any None values
        impairment_degree = np.mean(critical_levels) if len(critical_levels) > 0 else 0

        if impairment_degree > 56:
            return "SEVERE"
        if impairment_degree > 35:
            return "MODERATE"
        if impairment_degree > 15:
            return "MILD"

        return "NOTHING"

    def resample(self, new_frequencies: ndarray) -> Audiogram:
        """Resample the audiogram to a new set of frequencies.

        Args:
            new_frequencies (ndarray): The new frequencies to resample to

        Returns:
            Audiogram: New audiogram with resampled frequencies

        """
        return Audiogram(
            levels=logx_interp1d(self.frequencies, self.levels, new_frequencies),
            frequencies=new_frequencies,
        )


# Reference processing: use to check levels between original and processed,
AUDIOGRAM_REF = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
)

# mild age-related hearing loss, slightly reduced from CF, first-time aid wearers group
# (used in Stafa talk by MAS)
AUDIOGRAM_MILD = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([5, 10, 15, 18, 19, 22, 25, 28, 31, 35, 38, 40, 40, 45, 50]),
)

# mod hearing loss based on mild N2 flat/mod sloping from Bisgaard et al. 2020
AUDIOGRAM_MODERATE = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([15, 20, 20, 22.5, 25, 30, 35, 40, 45, 50, 55, 55, 60, 65, 65]),
)

# (mod-)severe age-related hearing loss, average of MAS/KA summer proj 2011, elderly HI
AUDIOGRAM_MODERATE_SEVERE = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([19, 19, 28, 35, 40, 47, 52, 56, 58, 58, 63, 70, 75, 80, 80]),
)
