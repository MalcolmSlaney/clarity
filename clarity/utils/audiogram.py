"""Dataclass to represent a monaural audiogram"""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Final

import numpy as np

if TYPE_CHECKING:
    from numpy import ndarray

DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES: Final = np.array(
    [
        250,
        500,
        1000,
        2000,
        3000,
        4000,
        6000,
        8000,
    ]
)

FULL_STANDARD_AUDIOGRAM_FREQUENCIES: Final = np.array(
    [
        125,
        250,
        500,
        750,
        1000,
        1500,
        2000,
        3000,
        4000,
        6000,
        8000,
        10000,
        12000,
        14000,
        16000,
    ]
)


@dataclass
class Audiogram:
    """Dataclass to represent an audiogram.

    Attributes:
        levels (ndarray): The audiometric levels in dB HL
        frequencies (ndarray): The frequencies at which the levels are measured

    """

    levels: np.ndarray
    frequencies: np.ndarray = np.array(DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES)

    def __post_init__(self) -> None:
        """Check that dimensions of levels and frequencies match."""
        if len(self.levels) != len(self.frequencies):
            raise ValueError(
                f"Levels ({len(self.levels)}) and frequencies ({len(self.frequencies)})"
                " must have the same length"
            )

    @property
    def severity(self) -> str:
        """Categorise HL severity level for the audiogram.

        Note that this categorisation is different from that of the British
        Society of Audiology, which recommends descriptors mild, moderate,
        severe and profound for average hearing threshold levels at 250, 500,
        1000, 2000 and 4000 Hz of 21-40 dB HL, 41-70 dB HL, 71-95 dB HL
        and > 95 dB HL, respectively (BSA Pure-tone air-conduction and
        bone-conduction threshold audiometry with and without masking
        2018).

        Returns:
            str -- severity level, one of SEVERE, MODERATE, MILD, NOTHING

        """
        # calculate mean hearing loss between critical frequencies of 2 & 8 kHz
        critical_freqs = np.logical_and(
            2000 <= self.frequencies, self.frequencies <= 8000
        )
        critical_levels = self.levels[critical_freqs]
        # Remove any None values
        critical_levels = [x for x in critical_levels if x is not None]
        # Ignore any None values
        impairment_degree = np.mean(critical_levels) if len(critical_levels) > 0 else 0

        if impairment_degree > 56.0:
            return "SEVERE"
        if impairment_degree > 35.0:
            return "MODERATE"
        if impairment_degree > 15.0:
            return "MILD"

        return "NOTHING"

    def resample(self, new_frequencies: ndarray) -> Audiogram:
        """Resample the audiogram to a new set of frequencies.

        Interpolates linearly on a log frequency axis.

        Args:
            new_frequencies (ndarray): The new frequencies to resample to

        Returns:
            Audiogram: New audiogram with resampled frequencies

        """

        return Audiogram(
            levels=np.interp(
                np.log(new_frequencies),
                np.log(self.frequencies),
                self.levels,
                left=self.levels[0],
                right=self.levels[-1],
            ),
            frequencies=new_frequencies,
        )


# Reference audiograms originally defined in the Cambridge group
# MSBG model MATLAB code.

# No loss
AUDIOGRAM_REF: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
)

# Mild age-related hearing loss, slightly reduced from CF, first-time aid
# wearers group (used in Stafa talk by MAS).
AUDIOGRAM_MILD: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([5, 10, 15, 18, 19, 22, 25, 28, 31, 35, 38, 40, 40, 45, 50]),
)

# Moderate hearing loss based on mild N2 flat/mod sloping
# from Bisgaard et al. 2020
AUDIOGRAM_MODERATE: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([15, 20, 20, 22.5, 25, 30, 35, 40, 45, 50, 55, 55, 60, 65, 65]),
)

# Moderate-severe age-related hearing loss, average of MAS/KA summer proj 2011,
# elderly HI
AUDIOGRAM_MODERATE_SEVERE: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([19, 19, 28, 35, 40, 47, 52, 56, 58, 58, 63, 70, 75, 80, 80]),
)
