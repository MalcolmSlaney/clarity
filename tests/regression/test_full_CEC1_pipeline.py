# Regression test
# Pass some random data through code and compare with reference output
# scene_renderer, enhancer, compressor, haspi

import os

import numpy as np
from omegaconf import OmegaConf
from scipy.io import wavfile
from scipy.signal import unit_impulse

from clarity.data.scene_renderer_cec2 import SceneRenderer
from clarity.enhancer.compressor import Compressor
from clarity.enhancer.nalr import NALR
from clarity.evaluator.mbstoi.mbstoi import mbstoi
from clarity.evaluator.mbstoi.mbstoi_utils import find_delay_impulse
from clarity.evaluator.msbg.audiogram import Audiogram
from clarity.evaluator.msbg.msbg import Ear
from clarity.evaluator.msbg.msbg_utils import MSBG_FS, pad


def listen(ear, signal, audiogram_l, audiogram_r):
    """
    Generate MSBG processed signal
    :param ear: MSBG ear
    :param wav: binaural signal
    :return: binaural signal
    """
    ear.set_audiogram(audiogram_l)
    out_l = ear.process(signal[:, 0])

    ear.set_audiogram(audiogram_r)
    out_r = ear.process(signal[:, 1])
    if len(out_l[0]) != len(out_r[0]):
        diff = len(out_l[0]) - len(out_r[0])
        if diff > 0:
            out_r[0] = np.flipud(pad(np.flipud(out_r[0]), len(out_l[0])))
        else:
            out_l[0] = np.flipud(pad(np.flipud(out_l[0]), len(out_r[0])))
    return np.concatenate([out_l, out_r]).T


def test_full_CEC1_pipeline(regtest):
    np.random.seed(0)

    # Set up some scene to simulate
    # It's been designed to get good code coverage but running quickly
    # - Using three maskers - one from each noise type
    # - Using a short target with reduce pre and post silence
    # - Only generating 2 hearing aid channels
    scene = {
        "dataset": "demo",
        "room": "R06001",
        "scene": "S06001",
        "target": {"name": "T010_G0N_02468", "time_start": 37837, "time_end": 115894},
        "duration": 150000,
        "interferers": [
            {
                "position": 1,
                "time_start": 0,
                "time_end": 150000,
                "type": "noise",
                "name": "CIN_fan_014.wav",
                "offset": 5376,
            },
            {
                "position": 2,
                "time_start": 0,
                "time_end": 150000,
                "type": "speech",
                "name": "som_04766_05.wav",
                "offset": 40000,
            },
            {
                "position": 3,
                "time_start": 0,
                "time_end": 150000,
                "type": "music",
                "name": "1111967.low.mp3",
                "offset": 842553,
            },
        ],
        "SNR": 0.0,
        "listener": {
            "rotation": [
                {"sample": 116192.9795, "angle": 52.3628},
                {"sample": 124829.9795, "angle": 38.5256},
            ],
            "hrir_filename": ["VP_N6-ED", "VP_N6-BTE_fr"],
        },
    }

    demo_paths = OmegaConf.create(
        {
            "hoairs": "tests/test_data/rooms/HOA_IRs",
            "hrirs": "tests/test_data/hrir/HRIRs_MAT",
            "scenes": "tests/test_data/clarity_data/demo/scenes",
            "targets": "tests/test_data/targets",
            "interferers": "tests/test_data/interferers/{type}",
        }
    )

    demo_metadata = OmegaConf.create(
        {
            "room_definitions": "tests/test_data/metadata/rooms.demo.json",
            "scene_definitions": "",  # Scene definition file not needed for test
            "hrir_metadata": "tests/test_data/metadata/hrir_data.json",
        }
    )

    scene_renderer = SceneRenderer(
        demo_paths,
        demo_metadata,
        ambisonic_order=6,
        equalise_loudness=True,
        reference_channel=1,
        channel_norms=[12.0, 3.0],
    )

    output_path = "tmp"

    if not os.path.exists(output_path):
        os.makedirs(output_path)

    target, interferers, anechoic, head_turn = scene_renderer.generate_hoa_signals(
        scene
    )

    scene_renderer.generate_binaural_signals(
        scene, target, interferers, anechoic, output_path
    )

    _, reference = wavfile.read(f"{output_path}/S06001_target_anechoic_CH1.wav")
    _, signal = wavfile.read(f"{output_path}/S06001_mix_CH1.wav")
    reference = reference.astype(float) / 32768.0
    signal = signal.astype(float) / 32768.0

    # Truncate to just over 2 seconds - i.e. just use part of the signals to speed up the HASPI calculation a little
    signal = signal[:100000, :]
    reference = reference[:100000, :]

    # The data below doesn't really need to be meaningful.
    # The purpose of the test is not to see if the haspi score is reasonable
    # but just to check that the results do not change unexpectedly across releases.

    nalr_cfg = {"nfir": 220, "fs": 44100}
    compressor_cfg = {
        "threshold": 0.35,
        "attenuation": 0.1,
        "attack": 50,
        "release": 1000,
        "rms_buffer_size": 0.064,
    }

    fs = 44100

    msbg_ear_cfg = {"src_pos": "ff", "fs": fs, "equiv0dBSPL": 100, "ahr": 20}

    audiogram_l = np.array([45, 50, 60, 65, 60, 65, 70, 80])
    audiogram_r = np.array([45, 45, 60, 70, 60, 60, 80, 80])
    audiogram_cfs = np.array([250, 500, 1000, 2000, 3000, 4000, 6000, 8000])

    ear = Ear(**msbg_ear_cfg)

    left_audiogram = Audiogram(cfs=audiogram_cfs, levels=audiogram_l)
    right_audiogram = Audiogram(cfs=audiogram_cfs, levels=audiogram_r)

    enhancer = NALR(**nalr_cfg)
    compressor = Compressor(**compressor_cfg)

    nalr_fir, _ = enhancer.build(audiogram_l, audiogram_cfs)
    out_l = enhancer.apply(nalr_fir, signal[:, 0])

    nalr_fir, _ = enhancer.build(audiogram_r, audiogram_cfs)
    out_r = enhancer.apply(nalr_fir, signal[:, 1])

    out_l, _, _ = compressor.process(out_l)
    out_r, _, _ = compressor.process(out_r)

    enhanced_audio = np.stack([out_l, out_r], axis=1)

    enhanced_audio = np.tanh(enhanced_audio)

    # Create discrete delta function (DDF) signal for time alignment
    ddf_signal = np.zeros((np.shape(signal)))
    ddf_signal[:, 0] = unit_impulse(len(signal), int(MSBG_FS / 2))
    ddf_signal[:, 1] = unit_impulse(len(signal), int(MSBG_FS / 2))

    reference_processed = listen(ear, reference, left_audiogram, right_audiogram)
    signal_processed = listen(ear, signal, left_audiogram, right_audiogram)

    # Calculate channel-specific unit impulse delay due to HL model and audiograms
    delay = find_delay_impulse(ddf_signal, initial_value=int(MSBG_FS / 2))
    max_delay = int(np.max(delay))

    # Allow for value lower than 1000 samples in case of unimpaired hearing
    assert max_delay <= 2000

    # Correct for delays by padding clean signals
    clean_pad = np.zeros((len(reference_processed) + max_delay, 2))
    proc_pad = np.zeros((len(signal_processed) + max_delay, 2))

    assert len(proc_pad) >= len(signal_processed)

    clean_pad[
        int(delay[0]) : int(len(reference_processed) + int(delay[0])), 0
    ] = reference_processed[:, 0]
    clean_pad[
        int(delay[1]) : int(len(reference_processed) + int(delay[1])), 1
    ] = reference_processed[:, 1]
    proc_pad[: len(signal_processed)] = signal_processed

    grid_coarseness = 1
    sii_enhanced = mbstoi(
        clean_pad[:, 0],
        clean_pad[:, 1],
        proc_pad[:, 0],
        proc_pad[:, 1],
        44100,
        grid_coarseness,
    )

    print(f"Enhanced audio MBSTOI score is {sii_enhanced}")

    regtest.write(f"Enhanced audio MBSTOI score is {sii_enhanced:0.7f}\n")

    # Enhanced audio MBSTOI score is 0.2994066
