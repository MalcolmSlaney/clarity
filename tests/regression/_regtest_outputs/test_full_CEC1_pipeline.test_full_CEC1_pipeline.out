Enhanced audio MBSTOI score is 0.5204395
